File: backends/go-gin/internal/presentation/middleware.go
package presentation

import (
	"net/http"
	"strings"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"

	"url-shortener-go/internal/domain"
	"url-shortener-go/internal/infrastructure"
)

const ContextUserKey = "user"

type AuthUser struct {
	UserID   uuid.UUID
	TenantID uuid.UUID
	Email    string
}

func AuthMiddleware(cfg *infrastructure.Config) gin.HandlerFunc {
	return func(c *gin.Context) {
		authHeader := c.GetHeader("Authorization")
		if authHeader == "" {
			c.JSON(http.StatusUnauthorized, gin.H{"error": "Missing authorization header"})
			c.Abort()
			return
		}

		parts := strings.SplitN(authHeader, " ", 2)
		if len(parts) != 2 || strings.ToLower(parts[0]) != "bearer" {
			c.JSON(http.StatusUnauthorized, gin.H{"error": "Invalid authorization header"})
			c.Abort()
			return
		}

		claims, err := infrastructure.VerifyToken(parts[1], cfg)
		if err != nil {
			c.JSON(http.StatusUnauthorized, gin.H{"error": "Invalid or expired token"})
			c.Abort()
			return
		}

		if claims.Type != "access" {
			c.JSON(http.StatusUnauthorized, gin.H{"error": "Invalid token type"})
			c.Abort()
			return
		}

		userID, err := uuid.Parse(claims.Subject)
		if err != nil {
			c.JSON(http.StatusUnauthorized, gin.H{"error": "Invalid user ID"})
			c.Abort()
			return
		}

		tenantID, err := uuid.Parse(claims.TenantID)
		if err != nil {
			c.JSON(http.StatusUnauthorized, gin.H{"error": "Invalid tenant ID"})
			c.Abort()
			return
		}

		c.Set(ContextUserKey, &AuthUser{
			UserID:   userID,
			TenantID: tenantID,
			Email:    claims.Email,
		})
		c.Next()
	}
}

func GetAuthUser(c *gin.Context) (*AuthUser, *domain.AppError) {
	val, exists := c.Get(ContextUserKey)
	if !exists {
		return nil, domain.NewUnauthorizedError("User not found in context")
	}
	user, ok := val.(*AuthUser)
	if !ok {
		return nil, domain.NewInternalError()
	}
	return user, nil
}

File: backends/rust-axum/src/presentation/middleware.rs
use axum::{
    extract::Request,
    http::{header::AUTHORIZATION, StatusCode},
    middleware::Next,
    response::Response,
};
use uuid::Uuid;

use crate::domain::errors::AppError;
use crate::infrastructure::{auth, config::Settings};

#[derive(Debug, Clone)]
pub struct AuthUser {
    pub user_id: Uuid,
    pub tenant_id: Uuid,
    pub email: String,
}

pub async fn auth_middleware(
    settings: axum::extract::State<Settings>,
    request: Request,
    next: Next,
) -> Result<Response, AppError> {
    let auth_header = request
        .headers()
        .get(AUTHORIZATION)
        .ok_or_else(|| AppError::Unauthorized("Missing authorization header".to_string()))?;

    let auth_str = auth_header
        .to_str()
        .map_err(|_| AppError::Unauthorized("Invalid authorization header".to_string()))?;

    if !auth_str.starts_with("Bearer ") {
        return Err(AppError::Unauthorized("Invalid authorization header".to_string()));
    }

    let token = &auth_str[7..];
    let claims = auth::verify_token(token, &settings.jwt)
        .map_err(|_| AppError::Unauthorized("Invalid or expired token".to_string()))?;

    if claims.token_type != "access" {
        return Err(AppError::Unauthorized("Invalid token type".to_string()));
    }

    let user = AuthUser {
        user_id: Uuid::parse_str(&claims.sub)
            .map_err(|_| AppError::Unauthorized("Invalid user ID".to_string()))?,
        tenant_id: Uuid::parse_str(&claims.tenant_id)
            .map_err(|_| AppError::Unauthorized("Invalid tenant ID".to_string()))?,
        email: claims.email,
    };

    let mut request = request;
    request.extensions_mut().insert(user);

    Ok(next.run(request).await)
}
